src/next/middleware.ts

```diff
@@ -0,0 +1,162 @@
+) {
+  const options = optionsOrEv as NextAuthMiddlewareOptions // This is relatively safe as we don't share properties with NextFetchEvent
+  const signInPage = options?.pages?.signIn ?? "/api/auth/signin"
+  const errorPage = options?.pages?.error ?? "/api/auth/error"
+
+  // Avoid infinite redirect loop
+  if ([signInPage, errorPage].includes(req.nextUrl.pathname)) return
+
+  const secret = options?.secret ?? process.env.NEXTAUTH_SECRET
+  // Continue only if the secret is specified
+  if (secret)
+    return {
+      req,
+      options,
+      secret,
+      signInPage,
+      authorized: options?.authorized ?? (({ token }) => token),
+    }
+
+  console.error(
+    `[next-auth][error][NO_SECRET]`,
+    `\nhttps://next-auth.js.org/errors#no_secret`
+  )
+
+  return {
+    redirect: NextResponse.redirect(`${errorPage}?error=Configuration`),
+  }
+}
+
+/**
+ * Middleware that checks if the user is authenticated/authorized.
+ * If if they aren't, they will be redirected to the login page.
+ * Otherwise, continue.
+ *
+ * @example
+ *
+ * ```js
+ * // `pages/_middleware.js`
+ * export { default } from "next-auth/middleware"
+ * ```
+ *
+ * ---
+ * [Documentation](https://next-auth.js.org/getting-started/middleware)
+ */
+export function withAuth(...args: WithAuthArgs) {
+  if (args[0] instanceof NextRequest) {
+    const config = initConfig(args[0], args[1])
+    if (!config || config.redirect) return config?.redirect
+    const { req, secret, signInPage, authorized } = config
+
+    return getToken({ req: req as any, secret }).then(async (token) => {
+      if (await authorized({ req, token })) return
+
+      return NextResponse.redirect(
+        `${signInPage}?${new URLSearchParams({ callbackUrl: req.url })}`
+      )
+    })
+  }
+
+  if (typeof args[0] === "function") {
+    const middleware = args[0]
+    const options = args[1] as NextAuthMiddlewareOptions | undefined
+    return async (...args: Parameters<NextMiddleware>) => {
+      const config = initConfig(args[0], options)
+      if (!config || config.redirect) return config?.redirect
+      const { req, secret, signInPage, authorized } = config
+
+      const token = await getToken({ req: req as any, secret })
+
+      if (await authorized({ req, token })) {
+        ;(args[0] as any).token = token
+        return await middleware(...args)
+      }
+
+      return NextResponse.redirect(
+        `${signInPage}?${new URLSearchParams({ callbackUrl: req.url })}`
+      )
+    }
+  }
+
+  const options = args[0]
+  return async (...args: Parameters<NextMiddleware>) => {
+    const config = initConfig(args[0], options)
+    if (!config || config.redirect) return config?.redirect
+
+    const { req, secret, signInPage, authorized } = config
+
+    const token = await getToken({ req: req as any, secret })
+
+    if (await authorized({ req, token })) return
+
+    return NextResponse.redirect(
+      `${signInPage}?${new URLSearchParams({ callbackUrl: req.url })}`
+    )
+  }
+}
```

Comment:
I wonder if abstracting the request handling flow into a **helper function** would help in this case 🤔 :
```ts
async function handleRequest(config, onSuccess = () => {}) {
  // if there was an error reading the middleware config redirect to the error page
  if (config.errorRedirect)  return config.errorRedirect
  
  // prevent an infinite loop when coming from an OAuth flow like `/api/auth/callback  /{provider}`
  if (!config) return
  
  const { req, secret, signInPage, authorized } = config
  
  const token = await getToken({ req: req as any, secret })
  const isAuthorized = await authorized({ req, token })
  
  // the user is authorized, let the middleware handle the rest
  if (isAuthorized) return onSuccess()
  
  // the user is not logged in, re-direct to the sign-in page
  return NextResponse.redirect(
    `${config.signInPage}?${new URLSearchParams({ callbackUrl: req.url })}`
  )
}
```
It seems that for the **3 cases**:
- `args[0] instanceof NextRequest` 
- `typeof args[0] === "function"`
- `args[0] === options`

The flow is the same apart from what happens on success, right? 🤔

Reply:
You were right! I did refactor everything according to this.